kubectl-plugin/pkg/cmd/job/job_submit.go

```diff
@@ -0,0 +1,565 @@
+	if rayJobAnnotations == nil {
+		rayJobAnnotations = make(map[string]string)
+	}
+
+	rayJobAnnotations["ray.io/ray-job-submission-id"] = rayJobID
+	options.RayJob.SetAnnotations(rayJobAnnotations)
+
+	_, err = k8sClients.DynamicClient().Resource(util.RayJobGVR).Namespace(*options.configFlags.Namespace).Update(ctx, options.RayJob, v1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("Error occurred when trying to add job ID to rayJob: %w", err)
+	}
+
+	// Wait for ray job submit to finish.
+	err = cmd.Wait()
+	if err != nil {
+		return fmt.Errorf("Error occurred with ray job submit: %w", err)
+	}
+	return nil
+}
+
+func (options *SubmitJobOptions) raySubmitCmd() ([]string, error) {
+	raySubmitCmd := []string{"ray", "job", "submit", "--address", dashboardAddr}
+
+	if len(options.runtimeEnv) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--runtime-env", options.runtimeEnv)
+	}
+	if len(options.runtimeEnvJson) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--runtime-env-json", options.runtimeEnvJson)
+	}
+	if len(options.submissionID) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--submission-id", options.submissionID)
+	}
+	if options.entryPointCPU > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--entrypoint-num-cpus", fmt.Sprintf("%f", options.entryPointCPU))
+	}
+	if options.entryPointGPU > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--entrypoint-num-gpus", fmt.Sprintf("%f", options.entryPointGPU))
+	}
+	if options.entryPointMemory > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--entrypoint-memory", fmt.Sprintf("%d", options.entryPointMemory))
+	}
+	if len(options.entryPointResource) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--entrypoint-resource", options.entryPointResource)
+	}
+	if len(options.metadataJson) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--metadata-json", options.metadataJson)
+	}
+	if options.noWait {
+		raySubmitCmd = append(raySubmitCmd, "--no-wait")
+	}
+	if len(options.headers) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--headers", options.headers)
+	}
+	if len(options.verify) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--verify", options.verify)
+	}
+	if len(options.logStyle) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--log-style", options.logStyle)
+	}
+	if len(options.logColor) > 0 {
+		raySubmitCmd = append(raySubmitCmd, "--log-color", options.logColor)
+	}
+
+	raySubmitCmd = append(raySubmitCmd, "--working-dir", options.workingDir)
+
+	raySubmitCmd = append(raySubmitCmd, "--")
+	// Sanitize entrypoint
+	entryPointSanitized, err := shlex.Split(options.entryPoint)
+	if err != nil {
+		return nil, err
+	}
+	raySubmitCmd = append(raySubmitCmd, entryPointSanitized...)
+
+	return raySubmitCmd, nil
+}
+
+// Decode rayjob yaml if we decide to submit job using kube client
+func decodeRayJobYaml(rayJobFilePath string) (*unstructured.Unstructured, error) {
+	openRayJobYamlFile, err := os.Open(rayJobFilePath)
+	if err != nil {
+		return nil, err
+	}
+	defer openRayJobYamlFile.Close()
```

Comment:
There lines are redundant.

Reply:
Can you explain what the redundancy is here? I don't completely understand. Is it the deferred Close() immediately after the error check thats causing the redundancy?

Reply:
This variable `openRayJobYamlFile` is not used.

Reply:
Sorry, my English expression isn’t very precise. Maybe I need to say this variable is not used instead of "redundant" in the beginning.

Reply:
Oh no, you're fine, I should've noticed it when I switched to `os.ReadFile`. That's my fault, thanks for catching it.